### PR TITLE
fix(react-router): remove flashes caused by `<ScrollRestoration />`

### LINKED ISSUE #2601: `<ScrollRestoration />` does not restore the scroll positions before DOM paint
### Which project does this relate to?

Router

### Describe the bug

It shows the top of the page then restore the scroll positions. This causes flickering.

### Your Example Website or App

https://stackblitz.com/edit/tanstack-router-k6ncqr?file=src%2Fmain.tsx

### Steps to Reproduce the Bug or Issue

1. At the `/` route (**Home** tab), scroll to the bottom of the home tab.
2. Click the **About** tab.
3. Click the **Home** tab.
4. I can see the red background of the **Home** tab title appear and then suddenly disappeared.

### Expected behavior

No red background flickering

### Screenshots or Videos

https://github.com/user-attachments/assets/22f05924-bb58-4894-9938-5a7903b51db4

### Platform

- OS: WInd
- Browser: Edge
- Version: 130.0.2849.46

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by xiaohai-huang**
This PR includes the following 2 main changes to resolve #2601:

- Make the router emit a new event called `onBeforeRouteMount`.
- `<ScrollRestoration />` performs the restoration `onBeforeRouteMount` instead of `onResolved` to avoid flashes.

The `onBeforeRouteMount` is designed to trigger actions that should occur just before a new route mounts. This event allows developers to perform pre-render operations, such as scroll restoration, ensuring that these actions complete before the route content becomes visible to the user.

## Current Problems

I cannot get the tests to pass as it behaves differently from the browser environment. It works fine on browsers.

### CODE REVIEW

**Comment by xiaohai-huang** [comment]
@schiller-manuel Can you help me with the tests?

**Comment by schiller-manuel** [comment]
where exactly do you need help? do the tests fail? or are you unsure if the tests test what they should?

**Comment by schiller-manuel** [comment]
you could try not using `createMemoryHistory` instead use the default history.
If this still does not behave as a browser would, then those tests should be moved into the `e2e` tests where they are ran in a real browser.

File: packages/react-router/src/router.ts [hunk 1]
```diff
@@ -608,6 +608,12 @@ export type RouterEvents = {
     toLocation: ParsedLocation
     pathChanged: boolean
   }
+  onBeforePageMount: {
```
Comment (schiller-manuel) [inline packages/react-router/src/router.ts:611]: please also update docs/framework/react/api/router/RouterEventsType.md

File: packages/react-router/tests/router.test.tsx [hunk 2]
```diff
@@ -521,6 +521,142 @@ describe('router emits events during rendering', () => {
...
+    } else {
+      throw new Error(
+        '"onBeforePageMount" should be emitted before page\'s useLayoutEffect setup',
+      )
+    }
+
+    unsubBeforeMount()
+  })
+
+  it('after a navigation, should have emitted the "onBeforePageMount" event twice', async () => {
```
Comment (schiller-manuel) [inline packages/react-router/tests/router.test.tsx:624]: why twice?

**Comment by schiller-manuel** [comment]
this looks good, can we merge?

**Comment by xiaohai-huang** [comment]
> this looks good, can we merge?

Yes, please

**Pull request merged**